src/nwb_conversion_tools/tools/hdmf.py

```diff
@@ -90,3 +91,66 @@ def _get_maxshape(self) -> tuple:
 
     def _get_data(self, selection: Tuple[slice]) -> np.ndarray:
         return self.data[selection]
+
+
+class ImagingExtractorDataChunkIterator(GenericDataChunkIterator):
+    """
+    Generic data chunk iterator for an ImagingExtractor object
+    primarily used when writing imaging data to an NWB file.
+    """
+
+    def __init__(
+        self,
+        imaging_extractor: ImagingExtractor,
+        buffer_gb: Optional[float] = None,
+        buffer_shape: Optional[tuple] = None,
+        chunk_mb: Optional[float] = None,
+        chunk_shape: Optional[tuple] = None,
+        display_progress: bool = False,
+        progress_bar_options: Optional[dict] = None,
+    ):
+        self.imaging_extractor = imaging_extractor
+
+        if buffer_gb is None and buffer_shape is None:
+            buffer_gb = 1.0
+
+        if buffer_shape is None:
+            buffer_shape = self._get_default_buffer_shape(buffer_gb=buffer_gb)
+
+        if chunk_mb is None and chunk_shape is None:
+            chunk_mb = 1.0
+
+        super().__init__(
+            buffer_shape=buffer_shape,
+            chunk_mb=chunk_mb,
+            chunk_shape=chunk_shape,
+            display_progress=display_progress,
+            progress_bar_options=progress_bar_options,
+        )
+
+        assert self.buffer_shape[1:] == self.maxshape[1:], (
+            f"Except from the first axis, the buffer shape ({self.buffer_shape}) "
+            f"must be equal to max shape ({self.maxshape})."
+        )
```

Comment:
Assertions like this would typically go at the start so we can avoid triggering anything else as early as possible. You'd just have to get the `maxshape` for comparison to a manually set `buffer_shape` (and only in the case where the `buffer_shape` is set manually to begin with)